NPE: when  the "Holder of Key" as the subject confirmation method based on the WS-Trust

**Describe the issue:**
NPE has occurred.
If the "Holder of Key" as the subject confirmation method [1] based on the WS-Trust request which is received to WSO2.
This issue is fixed already in the master branch[2] latest versions as well. However for the Holder of key needs to improve the fix further as well.

```
TID: [-1234] [wso2carbon-sts] [2020-05-16 18:21:29,792] [27273ea1-0df5-4018-ae8e-3cfd9b0e8b70] ERROR {org.apache.rahas.STSMessageReceiver} - Error in creating a SAMLToken using Opensaml library org.apache.rahas.TrustException: Error in creating a SAMLToken using Opensaml library
        at org.apache.rahas.impl.SAMLTokenIssuer.createAttributeAssertion(SAMLTokenIssuer.java:629)
        at org.apache.rahas.impl.SAMLTokenIssuer.createHoKAssertion(SAMLTokenIssuer.java:458)
        at org.apache.rahas.impl.SAMLTokenIssuer.issue(SAMLTokenIssuer.java:203)
        at org.apache.rahas.TokenRequestDispatcher.handle(TokenRequestDispatcher.java:69)
        at org.apache.rahas.STSMessageReceiver.invokeBusinessLogic(STSMessageReceiver.java:56)
        at org.apache.axis2.receivers.AbstractInOutMessageReceiver.invokeBusinessLogic(AbstractInOutMessageReceiver.java:40)
        at org.apache.axis2.receivers.AbstractMessageReceiver.receive(AbstractMessageReceiver.java:110)
        at org.apache.axis2.engine.AxisEngine.receive(AxisEngine.java:180)
        at org.apache.axis2.transport.http.HTTPTransportUtils.processHTTPPostRequest(HTTPTransportUtils.java:173)
        at org.apache.axis2.transport.http.AxisServlet.doPost(AxisServlet.java:148)
        at org.wso2.carbon.core.transports.CarbonServlet.doPost(CarbonServlet.java:238)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
        at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61)
        at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128)
        at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:68)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
        at org.wso2.carbon.tomcat.ext.servlet.DelegationServlet.service(DelegationServlet.java:68)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.owasp.csrfguard.CsrfGuardFilter.doFilter(CsrfGuardFilter.java:72)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:126)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.wso2.carbon.tomcat.ext.filter.CharacterSetFilter.doFilter(CharacterSetFilter.java:65)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:126)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:490)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
        at org.wso2.carbon.identity.context.rewrite.valve.TenantContextRewriteValve.invoke(TenantContextRewriteValve.java:80)
        at org.wso2.carbon.identity.authz.valve.AuthorizationValve.invoke(AuthorizationValve.java:100)
        at org.wso2.carbon.identity.auth.valve.AuthenticationValve.invoke(AuthenticationValve.java:74)
        at org.wso2.carbon.tomcat.ext.valves.CompositeValve.continueInvocation(CompositeValve.java:99)
        at org.wso2.carbon.tomcat.ext.valves.TomcatValveContainer.invokeValves(TomcatValveContainer.java:49)
        at org.wso2.carbon.tomcat.ext.valves.CompositeValve.invoke(CompositeValve.java:62)
        at org.wso2.carbon.tomcat.ext.valves.CarbonStuckThreadDetectionValve.invoke(CarbonStuckThreadDetectionValve.java:146)
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:678)
        at org.wso2.carbon.tomcat.ext.valves.CarbonContextCreatorValve.invoke(CarbonContextCreatorValve.java:57)
        at org.wso2.carbon.tomcat.ext.valves.RequestCorrelationIdValve.invoke(RequestCorrelationIdValve.java:116)
        at org.wso2.carbon.tomcat.ext.valves.RequestEncodingValve.invoke(RequestEncodingValve.java:49)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:853)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1587)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
        at org.wso2.carbon.identity.provider.AttributeCallbackHandler.handle(AttributeCallbackHandler.java:151)
        at org.apache.rahas.impl.SAMLTokenIssuer.createAttributeAssertion(SAMLTokenIssuer.java:583)
        ... 61 more

```

**How to reproduce:**
Configure as for the official documentation.

**Expected behavior:**
No NPE

**Environment information** (_Please complete the following information; remove  any unnecessary fields_) **:**
 - Product Version: observed in 5.9.0
 - OS: Official Ubuntu Docker image
 - User store: AD

---
### Optional Fields

**Related issues:**
[1] - https://github.com/wso2-extensions/identity-inbound-auth-openid/pull/66
Apart from the fix[1]  we need to add the below improvements as well.
patch diff for fixing the issue,
```
Index: components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java	(revision 2757ddd4282157175961ccff471a8d4036b970ca)
+++ components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java	(date 1589741041740)
@@ -135,6 +135,8 @@
                             applicationManagementService.getServiceProviderByClientId(endPointReference, "wstrust",
                                     tenantDomain);
                     ClaimMapping[] claimMappings = serviceProvider.getClaimConfig().getClaimMappings();
+                    UserStoreManager userStoreManager = IdentityTenantUtil.getRealm(null, userIdentifier).
+                            getUserStoreManager();
 
                     for (int i = 0; i < claimMappings.length; i++) {
                         String localClaimUri = claimMappings[i].getLocalClaim().getClaimUri();
@@ -148,14 +150,20 @@
                             if (StringUtils.isNotBlank(remoteClaimSuffixValue) &&
                                     StringUtils.isNotBlank(remoteClaimPrefixValue)) {
 
-                                if(!authenticatedUser.isFederatedUser()) {
-                                    if (log.isDebugEnabled()) {
-                                        log.debug("Loading claim values from local UserStore for user: "
-                                                + authenticatedUser.toString());
-                                    }
-                                    localClaimValue = IdentityProviderServiceComponent.getRealmService().
-                                            getBootstrapRealm().getUserStoreManager().
-                                            getUserClaimValue(userIdentifier, localClaimUri, "default");
+                                if (authenticatedUser != null) {
+                                    if (!authenticatedUser.isFederatedUser()) {
+                                        if (log.isDebugEnabled()) {
+                                            log.debug("Loading claim values from local UserStore for user: "
+                                                    + authenticatedUser.toString());
+                                        }
+                                        localClaimValue = IdentityProviderServiceComponent.getRealmService().
+                                                getBootstrapRealm().getUserStoreManager().
+                                                getUserClaimValue(userIdentifier, localClaimUri, "default");
+                                    }
+                                } else {
+                                    log.warn("Authenticated User Object is null unable to get the requested claims " +
+                                            "for user: " + userIdentifier);
+                                    localClaimValue = userStoreManager.getUserClaimValue(userIdentifier, localClaimUri, "default");
                                 }
 
                                 if (StringUtils.isEmpty(localClaimValue)) {
@@ -176,6 +184,8 @@
                     throw new SAMLException("Error while loading SP specific claims", e);
                 } catch (org.wso2.carbon.user.core.UserStoreException e) {
                     throw new SAMLException("Error while loading claims of the user", e);
+                } catch (IdentityException e) {
+                    throw new SAMLException("Error while loading claims of the user", e);
                 }
 
             }
@@ -431,4 +441,5 @@
         return new RequestedClaimData();
     }
 
+
 }
```

--- a/components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java
+++ b/components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java
@@ -64,6 +64,7 @@ import java.util.Map;
 import java.util.StringTokenizer;
 
 import static org.wso2.carbon.identity.application.authentication.framework.util.FrameworkConstants.AUTHENTICATED_USER;
+import static org.wso2.carbon.user.core.UserCoreConstants.DEFAULT_PROFILE;
 import static org.wso2.carbon.utils.multitenancy.MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
 
 public class AttributeCallbackHandler implements SAMLCallbackHandler {
@@ -151,7 +152,7 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
                                 if (isHandlerCalledFromWSTrustSTSFlow(attrCallback)) {
                                     localClaimValue = IdentityProviderServiceComponent.getRealmService().
                                             getBootstrapRealm().getUserStoreManager().
-                                            getUserClaimValue(userIdentifier, localClaimUri, "default");
+                                            getUserClaimValue(userIdentifier, localClaimUri, DEFAULT_PROFILE);
                                 } else if(!authenticatedUser.isFederatedUser()) {
                                     if (log.isDebugEnabled()) {
                                         log.debug("Loading claim values from local UserStore for user: "
@@ -159,7 +160,7 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
                                     }
                                     localClaimValue = IdentityProviderServiceComponent.getRealmService().
                                             getBootstrapRealm().getUserStoreManager().
-                                            getUserClaimValue(userIdentifier, localClaimUri, "default");
+                                            getUserClaimValue(userIdentifier, localClaimUri, DEFAULT_PROFILE);
                                 }
 
                                 if (StringUtils.isEmpty(localClaimValue)) {
@@ -439,11 +440,13 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
     protected RequestedClaimData getRequestedClaim() {
         return new RequestedClaimData();
     }
-    
+
     private boolean isHandlerCalledFromWSTrustSTSFlow(SAMLAttributeCallback attributeCallback) {
         
-        // Authenticated user property is properly set during a passive STS flow. It is not done in the WS Trust based
-        // flow.
+        /*
+        Authenticated user property is properly set during a passive STS flow. It is not done in the WS Trust based
+        flow.
+         */
         return !(attributeCallback.getData().getInMessageContext().getProperty(AUTHENTICATED_USER) instanceof
                 AuthenticatedUser);
     }

--- a/components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java
+++ b/components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java
@@ -147,8 +147,12 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
 
                             if (StringUtils.isNotBlank(remoteClaimSuffixValue) &&
                                     StringUtils.isNotBlank(remoteClaimPrefixValue)) {
-
-                                if(!authenticatedUser.isFederatedUser()) {
+                                // WS trust flow does not set the authenticated user property.
+                                if (isHandlerCalledFromWSTrustSTSFlow(attrCallback)) {
+                                    localClaimValue = IdentityProviderServiceComponent.getRealmService().
+                                            getBootstrapRealm().getUserStoreManager().
+                                            getUserClaimValue(userIdentifier, localClaimUri, "default");
+                                } else if(!authenticatedUser.isFederatedUser()) {
                                     if (log.isDebugEnabled()) {
                                         log.debug("Loading claim values from local UserStore for user: "
                                                 + authenticatedUser.toString());
@@ -177,7 +181,6 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
                 } catch (org.wso2.carbon.user.core.UserStoreException e) {
                     throw new SAMLException("Error while loading claims of the user", e);
                 }
-
             }
         }
     }
@@ -352,7 +355,8 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
         try {
             if (MapUtils.isEmpty(requestedClaimValues)) {
                 try {
-                    if (authenticatedUser == null) {
+                    // WS trust flow does not set the authenticated user property.
+                    if (isHandlerCalledFromWSTrustSTSFlow(callback)) {
                         connector = IdentityTenantUtil.getRealm(null, userId).
                                 getUserStoreManager();
                         mapValues = connector.getUserClaimValues(MultitenantUtils.getTenantAwareUsername(userId),
@@ -400,7 +404,7 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
                         } else {
                             nameSpace = claimData.getUri();
                             if (nameSpace.contains("/") && nameSpace.length() > (nameSpace.lastIndexOf("/") + 1)) {
-                                // Custom claim uri should be in a format of http(s)://nameSpace/name 
+                                // Custom claim uri should be in a format of http(s)://nameSpace/name
                                 name = nameSpace.substring(nameSpace.lastIndexOf("/") + 1);
                                 nameSpace = nameSpace.substring(0, nameSpace.lastIndexOf("/"));
                             } else {
@@ -435,5 +439,12 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
     protected RequestedClaimData getRequestedClaim() {
         return new RequestedClaimData();
     }
-
+    
+    private boolean isHandlerCalledFromWSTrustSTSFlow(SAMLAttributeCallback attributeCallback) {
+        
+        // Authenticated user property is properly set during a passive STS flow. It is not done in the WS Trust based
+        // flow.
+        return !(attributeCallback.getData().getInMessageContext().getProperty(AUTHENTICATED_USER) instanceof
+                AuthenticatedUser);
+    }
 }

--- a/components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java
+++ b/components/org.wso2.carbon.identity.provider/src/main/java/org/wso2/carbon/identity/provider/AttributeCallbackHandler.java
@@ -404,7 +404,7 @@ public class AttributeCallbackHandler implements SAMLCallbackHandler {
                         } else {
                             nameSpace = claimData.getUri();
                             if (nameSpace.contains("/") && nameSpace.length() > (nameSpace.lastIndexOf("/") + 1)) {
-                                // Custom claim uri should be in a format of http(s)://nameSpace/name
+                                // Custom claim uri should be in a format of http(s)://nameSpace/name.
                                 name = nameSpace.substring(nameSpace.lastIndexOf("/") + 1);
                                 nameSpace = nameSpace.substring(0, nameSpace.lastIndexOf("/"));
                             } else {

